libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -238,64 +243,166 @@ string IRGenerator::generateInitialAssignment(VariableDeclaration const& _varDec
 	{
-		ABIFunctions abiFunctions(m_evmVersion, m_context.revertStrings(), m_context.functionCollector());
-		unsigned paramVars = make_shared<TupleType>(constructor->functionType(false)->parameterTypes())->sizeOnStack();
+		string paramsString;
+		if (nextContract != nullptr && baseParams.find(nextContract) != baseParams.end())
+			paramsString = baseParams.at(nextContract);
+		for (auto const& param: baseParams)
+			if (param.first != nextContract)
+				paramsString += (paramsString.empty() ? "" : ", ") + param.second;
+		return paramsString;
+	};
+
+	string constructorsCode;
+	map<ContractDefinition const*, string> baseConstructorParams;
+	for (size_t i = 0; i < _contract.annotation().linearizedBaseContracts.size(); ++i)
+	{
+		ContractDefinition const* contract = _contract.annotation().linearizedBaseContracts[i];
+		baseConstructorParams.erase(contract);
+
+		constructorsCode += m_context.functionCollector().createFunction(constructorName(*contract), [&]() {
+			Whiskers construtorFunction(R"(
+				function <functionName>(<params><comma><baseParams>) {
+					<initBaseArguments>
+					<?hasNextConstructor>
+						<nextConstructor>(<nextParams>)
+					</hasNextConstructor>
+					<initStateVariables>
+					<?hasUserDefinedConstructor>
+						<constructorBody>
+					</hasUserDefinedConstructor>
+				}
+			)");
+			string params;
+			if (contract->constructor())
+				for (ASTPointer<VariableDeclaration> const& varDecl: contract->constructor()->parameters())
+					params += (params.empty() ? "" : ", ") + m_context.addLocalVariable(*varDecl).commaSeparatedList();
+			construtorFunction("params", params);
+			string baseParamsString = listAllParams(baseConstructorParams);
+			construtorFunction("baseParams", baseParamsString);
+			construtorFunction("comma", !params.empty() && !baseParamsString.empty() ? ", " : "");
+			construtorFunction("functionName", constructorName(*contract));
+			pair<string, map<ContractDefinition const*, string>> evaluatedArgs = evaluateConstructorArguments(*contract);
+			baseConstructorParams.insert(evaluatedArgs.second.begin(), evaluatedArgs.second.end());
+			construtorFunction("initBaseArguments", evaluatedArgs.first);
+			construtorFunction("hasNextConstructor", i < _contract.annotation().linearizedBaseContracts.size() - 1);
+			if (i < _contract.annotation().linearizedBaseContracts.size() - 1)
+			{
+				ContractDefinition const* nextContract = _contract.annotation().linearizedBaseContracts[i + 1];
+				construtorFunction("nextConstructor",constructorName(*nextContract));
+				construtorFunction("nextParams", listAllParams(baseConstructorParams, nextContract));
+			}
+			construtorFunction("initStateVariables", initStateVariables(*contract));
+			FunctionDefinition const* constructor = contract->constructor();
+			construtorFunction("hasUserDefinedConstructor", constructor);
+			if (constructor)
+				construtorFunction("constructorBody", generate(constructor->body()));
```

Comment:
```suggestion
				constructorFunction("constructorBody", generate(constructor->body()));
```

Reply:
Can this be a real yul function call?

I'm just thinking about function modifiers and if they might be easier to handle if constructor functions and regular functions are done in the same way.

Reply:
Hm, but yeah, maybe it's fine as it is.

Reply:
I was thinking to leave user defined constructor as is and generate it in same way as with normal functions and have implicit constructor for each contract, calling userdefined one

Reply:
I can quickly change this. It doesn't meter now, but with new features, maybe it can introduce some weird bugs